### PR TITLE
CorrIndex implementation for comparing decomposition outputs

### PR DESCRIPTION
**Pull request by hmbaghdassarian**
@aarmey  I have an implementation of CorrIndex here that can hopefully provide a good starting point

### CODE REVIEW

File: tensorly/metrics/similarity.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        absolute value of the product of two MxN matrices that have had their columns L2 normalized, 
+        taking the complex conjugate transpose of the first matrix -- C = |X1^H * X2|
+
+    Returns
+    -------
+    float
+        CorrIndex metric [0,1]; lower score means higher between matrices
+    """
+    
+    MN = C.shape[1]*C.shape[0]
```
Comment (earmingol) [inline tensorly/metrics/similarity.py:42]: Given that C is generated from C = np.abs(np.matmul(np.matrix(X_1).H, X_2)) (in Line 76) and both X_1, X_2 must be of shape M x N,  C should be of shape N x N. So C.shape[1] should be equal to C.shape[0]. Maybe include an assert condition before this.

File: tensorly/metrics/similarity.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    return (1/(MN))*(np.sum(np.abs(np.amax(C, 1) - 1)) + np.sum(np.abs(np.amax(C, 0) - 1)))
```
Comment (aarmey) [inline tensorly/metrics/similarity.py:43]: To make this work for other backends, you'll want to use the Tensorly operations. You can add `import tensorly as tl` at the top, then most of these can just be changed from `np.X` to `tl.X`. I think `amax` might have to become `max`.

File: tensorly/metrics/similarity.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+from typing import List
+
+import numpy as np
+from sklearn.preprocessing import normalize
```
Comment (aarmey) [inline tensorly/metrics/similarity.py:4]: Tensorly doesn't have sklearn as a dependency, so we can't use this. Luckily there's a `tl.norm` operation to get the norm for normalization.

File: tensorly/metrics/similarity.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    
+    if X_1.shape != X_2.shape: 
+        raise ValueError('Factor matrices should be of the same shapes')
+    
+    # normalize columns to L2 norm - even if ran decomposition with normalize_factors = True  
+#   X/np.apply_along_axis(np.linalg.norm, 0, X) 
+    X_1 = normalize(X_1, norm = 'l2', axis = 0)
+    X_2 = normalize(X_2, norm = 'l2', axis = 0)
+    
+    C = np.abs(np.matmul(np.matrix(X_1).H, X_2))
```
Comment (aarmey) [inline tensorly/metrics/similarity.py:76]: The element-wise complex conjugate is implemented as `tl.conj`, so this should be possible to do with that and a transpose.

**Review by aarmey — COMMENTED** [review]
A couple initial suggestions.

File: tensorly/metrics/similarity.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    Returns
+    -------
+    np.ndarray
+        A vertically stacked matrix of shape [sum(tensor.shape)xR)] 
+    """
+
+    if len({A.shape[1] for A in factors}) != 1:
+        raise ValueError('factors should be a list of loading matrices')
+    return tl.concatenate(factors)
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:26]: I don't think we need a function for that?

File: tensorly/metrics/similarity.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+def calc_corrindex(C: np.ndarray) -> float:
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:28]: What about just `correlation_index`?

File: tensorly/metrics/similarity.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    """Calculate the CorrIndex.
+
+    Parameters
+    ----------
+    C : np.ndarray
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:33]: Could we use a more explicit name for C?

File: tensorly/metrics/similarity.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    Returns
+    -------
+    float
+        CorrIndex metric [0,1]; lower score means higher between matrices
+    """
+    if C.shape[1] != C.shape[0]:
+        raise ValueError('C must be a square matrix')
+
+    MN = C.shape[1]*C.shape[0] #TODO: N notation in equation 8 unclear, make sure this is correct
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:45]: what about size or total size? Also better to use tl.shape(C)[1] just to make it backend agnostic.

File: tensorly/metrics/similarity.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    factors_2 : List[np.ndarray]
+        The loading/factor matrices [A^1 ... A^R] for a low-rank tensor from its factors, output from second decomposition
+    tol : float, optional
+        precision threshold below to call the CorrIndex score 0, by default 1e-16
+
+    Returns
+    -------
+    score : float
+         CorrIndex metric [0,1]; lower score means higher similarity between matrices
+    """
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:66]: I think this function and the one above can be merged and just called correlation_index, what do you think?

File: tensorly/metrics/similarity.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        precision threshold below to call the CorrIndex score 0, by default 1e-16
+
+    Returns
+    -------
+    score : float
+         CorrIndex metric [0,1]; lower score means higher similarity between matrices
+    """   
+    # get stacked loading matrices
+    X_1 = stack_loading_matrices(factors_1)
+    X_2 = stack_loading_matrices(factors_2)
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:69]: We can just do the concatenation here, no need for a separate function?

File: tensorly/metrics/similarity.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    if X_1.shape != X_2.shape:
+        raise ValueError('Factor matrices should be of the same shapes')
+
+    # normalize columns to L2 norm - even if ran decomposition with normalize_factors = True  
+    X_1 = X_1/tl.norm(X_1, axis = 0)
```
Comment (JeanKossaifi) [inline tensorly/metrics/similarity.py:75]: Shouldn't we check whether the norm is non-zero?
Also, we don't need spaces around `=` as it indicates a keyword argument (pep8)

**Review by JeanKossaifi — CHANGES_REQUESTED** [review]
Thanks @hmbaghdassarian and @aarmey!
I've left a few comments, mostly for clarity.

**Comment by hmbaghdassarian** [comment]
@JeanKossaifi thanks for the feedback, I tried implementing your suggestions. Let me know if I missed anything or if you have more changes you want

File: tensorly/metrics/similarity.py [hunk 12]
```diff
@@ -1,83 +1,55 @@
...
+    X_1 = X_1/col_norm_1
+    X_2 = X_2/col_norm_2
 
-    C = tl.abs(tl.matmul(tl.conj(X_1.T), X_2))
+    # generate the correlation index input
+    c_prod_mtx = tl.abs(tl.matmul(tl.conj(X_1.T), X_2))
 
-    score = calc_corrindex(C)
+    # correlation index scoring
+    n_elements = tl.shape(c_prod_mtx)[1] * tl.shape(c_prod_mtx)[0] #TODO: N notation in equation 8 unclear, make sure this is correct
```
Comment (earmingol) [inline tensorly/metrics/similarity.py:50]: This should be
```n_elements = tl.shape(c_prod_mtx)[1] + tl.shape(c_prod_mtx)[0]```
To meet the "2N" in the equation of the paper

Comment (aarmey) [reply tensorly/metrics/similarity.py]: Thanks! That's updated.

**Comment by aarmey** [comment]
I've now made some changes so that this *should* work for all backends.

**Comment by earmingol** [comment]
> 

Thanks @aarmey ! We have just one change to implement. It is an option for computing the index either on the stacked matrices or on a 1-to-1 manner on each pair of A matrices and then select the worse score.

**Comment by JeanKossaifi** [comment]
Looks good to me - @aarmey @cohenjer for an additional pair of eyes :)
Beautiful docstring for `correlation_index` 👌

**Review by aarmey — APPROVED** [review]
Also looks great to me! Looking forward to using this.
